src/components/alarm/AlarmController.cpp

```diff
@@ -102,6 +118,12 @@ void AlarmController::SetOffAlarmNow() {
 
 void AlarmController::StopAlerting() {
   systemTask->PushMessage(System::Messages::StopRinging);
+}
+
+void AlarmController::OnStopRinging() {
+  if (state != AlarmState::Alerting) {
+    return;
+  }
```

Comment:
The SystemTask message `StopRinging` is supposed to just stop the MotorController ringing. Combining the behaviour with the Alarm is confusing. Why was this change made?

Reply:
This postpones processing and makes sure that ScheduleAlarm is called for recurring alarms by the system task.

For reliability, in my modifications I tried to make sure that timers are started by the system task in response to system messages.

Reply:
I'm not sure SystemTask is more or less reliable in this case, but requiring it to go through another hoop and making the code less readable could be considered to make it less reliable.

Reply:
I have tried to enforce this discipline in my code because the FreeRTOS timer primitives are sensitive to their calling context, as they use mutexes and limited-size queues.
It is possible that this change is not necessary but it may require rather complex testing or analysis (stopAlerting currently has 5 call sites including event handlers, destructors and callbacks) and possibly more complex code changes.

Reply:
If there's concern that the queue can get full, the timer commands can be blocked for some time to let the timer task make room in the queue, or even read the return value, which is pdFAIL in case the command couldn't be sent. This is the intended and preferred way to handle this issue, however I wouldn't even be concerned about this right now, since the queue is length 32 and won't get full under normal circumstances.

Reply:
I see, the timer queue shouldn't get full if the timer task is awaken often enough and not accidentally blocked by a callback.
What about two tasks trying to access the timer queue simultaneously?

Reply:
https://freertos.org/Embedded-RTOS-Queues.html It says here that queues are used as thread safe buffers, so I'm pretty sure any concerns are handled in FreeRTOS.

Reply:
Indeed they are thread safe as they are protected by a mutex.  So what could happen is that one of the two tasks trying to access the queue in wait-free mode will fail.  (This may not happen if the access to the timer queue is further protected by a critical section to prevent preemption by the scheduler, but I have not been able to confirm that yet by looking at documentation or at FreeRTOS source code yet)

Reply:
Using blocking mode for timer primitives as you hinted before may not be an option in this case because StopAlerting is called by an event handler, among other contexts, where blocking may not be recommended. 

Recovery options in case of failure have their own drawbacks or problems, for example even though you can introduce a new system message to recover from a failed timer command, you will not be able to push it the system task because StopAlerting is already sending a message to the system task, and the system task queue has size 1.

I think it's possible that the timer queue can get full in some cases and cause such failures, for example in times of high cpu load which can cause the timer task to execute callbacks slowly and become congested, such a congestion may last only a few millisecond and not be noticeable by the user, but it can still be relevant to correct code execution.  Delaying the execution of the timer commands and letting the system task wake up again gives a better chance for the timer task to process events too and to become less congested.

Reply:
The documentation says this:
> xBlockTime | Specifies the time, in ticks, that the calling task should be held in the Blocked state to wait for the reset command to be successfully sent to the timer command queue, should the queue already be full when xTimerReset() was called.

There's no mention that the commands couldn't be called simultaneously because of accessing the queue. I'm sure it would have been mentioned somewhere if it was an issue. I don't see any reason to doubt FreeRTOS working as expected in this case. Without digging too deep into it, I can see that in `xQueueGenericSend()` in queue.c, it enters critical before doing anything.

SystemTask queue is length 10.

> in times of high cpu load which can cause the timer task to execute callbacks slowly and become congested

https://freertos.org/implementation/a00008.html I don't see how that could happen, but if there's still concern, let's move that to a new issue or PR, so we can move forward with this PR.

Reply:
Thank you for researching about FreeRTOS and correcting about the size of the system task, somehow I misread that in the source code,  About FreeRTOS that's great because it makes timers more reliable and easy to use in wait-free mode, as they are in the current codebase.
As far as I know nobody understood yet why the app_timer component was malfunctioning for alarms.  The implementation of app_timer uses FreeRTOS timer commands internally.  Perhaps the timer queue was getting full.  For averting such issues I think that trying to space out timer commands using the system message queue is a good discipline, and it's what worked in my tests while developing this solution.  However I also understand concerns about code clarity.  It's not easy to make asynchronous code clear, but I tried to keep code changes minimal.

Reply:
Since it's possible to send more than one message to the system task, I think I can try to develop a more clear solution than the current one, by introducing a RESCHEDULE_ALARM event.  This should also not increase the firmware size, compared to the present solution.  Do you think it's a good idea?

Reply:
That seems to explain the observed behaviour, it seemed that alarms were being disabled after ScheduleAlarm occurred when the alarm was already set, at least in some cases.

Higher priority for the timer task, compared to the task(s) sending timer commands, should also prevent in general the timer queue from getting full, as long as timer callbacks are not performing blocking operations.

Perhaps there is an existing rationale for the timer and the system task having the same priority?

If not, even if we may need to increase the number of allowed task priorities, it may become safe to reduce the length of the timer task queue, and hopefully achieve a slight reduction of RAM usage.

Reply:
I did another experiment where I read the return value of `app_timer_start()` and set off the alarm immediately depending on the error. If the queue is full, it returns `NRF_ERROR_NO_MEM`. If it thinks the timer is alrady running, it returns NRF_SUCCESS. When changing the time, the alarm wasn't set off, which proves that app_timer_start returned NRF_SUCCESS, queue was not full, yet it didn't alert.
```
-  app_timer_start(alarmAppTimer, APP_TIMER_TICKS(mSecToAlarm), this);
+  if (app_timer_start(alarmAppTimer, APP_TIMER_TICKS(mSecToAlarm), this) != NRF_SUCCESS) {
+    SetOffAlarmNow();
+  }
```

So with this in mind, please revert the extra changes from this PR and strictly focus on replacing app_timers. Let's also move the time adjustment change to a new PR. Thanks.